SONAR-6366 new ws for db migration

sns-seb opened the pull request:
this PR adds the new WS `server/migrate` 

IT tests coming up as soon as I know how they are supposed to be written

this PR is based on branch `feature/add_db_migration_from_java` reviewed with PR #197

jblievremont commented on server/sonar-server/src/main/java/org/sonar/server/platform/ws/ServerWsAction.java:
```diff
@@ -0,0 +1,29 @@
+import org.sonar.api.server.ws.RequestHandler;
+import org.sonar.api.server.ws.WebService;
+
+public interface ServerWsAction extends RequestHandler {
```
Extend ``WsAction`` instead (it extends ``Definable<WebService.NewController>``, which declares ``define``)

jblievremont commented on server/sonar-server/src/test/java/org/sonar/server/platform/ws/ServerMigrateWsActionTest.java:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    private static String expectedResponse(boolean operational, DatabaseMigration.Status status, String msg) {
+        return "{\"operational\":" + operational + ",\"state\":\"" + status + "\",\"message\":\"" + msg + "\"}";
```
You might want to extract those in JSON files and use WsTester.Result.assertJson() for readability (though it introduces an indirection).

jblievremont commented on server/sonar-server/src/test/java/org/sonar/server/platform/ws/ServerWsTest.java:
```diff
@@ -29,7 +31,7 @@
 public class ServerWsTest {
 
-  WsTester tester = new WsTester(new ServerWs());
+  WsTester tester = new WsTester(new ServerWs(new ServerWsAction[]{new DummyServerWsAction()}));
```
Not sure of the value of testing Pico injection :) instead you could inject a real ServerMigrateWsAction and check that it is correctly declared.

sns-seb replied:
This is not testing Pico injection, this is testing code looping over injected actions in `WsTester`
(and was actually useful since this a way I found that had forgotten to write the loop at first)

jblievremont commented:
Some small feedback, then LGTM :+1:

sns-seb commented:
rebased on updated #197 and took feedback into account

sns-seb commented:
rebased on updated #197 
+ fixed incorrect behavior caused by use of component `DefaultServerUpgradeStatus` which retrieved the current version of the DB *before* any upgrade could have been performed
+ fixed status text in JSON which did not match text from ROR WS
+ tested on local SQ instance, WS is responding

GithubIntegration commented:
SonarQube analysis reported 6 new issues:
* 1 blocking
* 4 major
* 1 minor

GithubIntegration commented:
SonarQube analysis reported 12 new issues:
* 1 blocking
* 7 major
* 2 minor
* 2 info

GithubIntegration commented:
SonarQube analysis reported 12 new issues:
* 1 blocking
* 7 major
* 2 minor
* 2 info

sns-seb commented:
note to self : JSON response should use a root object for extensibility

sns-seb commented:
PR rebased on PR #197 after some classes have been moved and renamed

GithubIntegration commented:
SonarQube analysis reported 16 new issues:
* 1 blocking
* 9 major
* 3 minor
* 3 info
